### PR TITLE
New composer: show format bar on selection

### PR DESCRIPTION
**Pull request by bwindels**
First PR for https://github.com/vector-im/riot-web/issues/10734

formatbar2

### CODE REVIEW

**Comment by bwindels** [comment]
Removed review request because detecting the selection doesn't work in Chrome for some reason.

File: src/components/views/rooms/BasicMessageComposer.js [hunk 1]
```diff
@@ -239,6 +247,36 @@ export default class BasicMessageEditor extends React.Component {
...
+
+                let topOffset = 0;
+                node = this._formatBarRef;
+                while (node.offsetParent) {
+                    node = node.offsetParent;
+                    topOffset += node.offsetTop;
+                }
+
+                this._formatBarRef.style.left = `${selectionRect.left - leftOffset}px`;
+                this._formatBarRef.style.top = `${selectionRect.top - topOffset - 16 - 12}px`;
```
Comment (turt2live) [inline src/components/views/rooms/BasicMessageComposer.js]: would be good to comment what the 16 and 12 do

File: src/components/views/rooms/BasicMessageComposer.js [hunk 2]
```diff
@@ -392,6 +430,42 @@ export default class BasicMessageEditor extends React.Component {
...
+            document.getSelection());
+        formatInline(range, prefix, suffix);
+    }
+
+    _formatBold = () => {
+        this._wrapSelectionAsInline("**");
+    }
+
+    _formatItalic = () => {
+        this._wrapSelectionAsInline("*");
```
Comment (turt2live) [inline src/components/views/rooms/BasicMessageComposer.js]: Should this not be toggling it off when the user already has it italicized? (same for others)

File: src/editor/operations.js [hunk 3]
```diff
@@ -0,0 +1,98 @@
+/*
+Copyright 2019 New Vector Ltd
```
Comment (turt2live) [inline src/editor/operations.js]: subtract this? I don't think we're claiming copyright on this

File: src/components/views/rooms/BasicMessageComposer.js (hunk 2 above)
Comment (bwindels) [reply src/components/views/rooms/BasicMessageComposer.js]: Ideally yes, but it's not part of the MVP and would be a bit trickier to do.

**Comment by bwindels** [comment]
Thanks for the review!

> I think this is generally sane - I'm getting a bit concerned about how much math we have just to deal with text though.

You mean the code inside `src/editor/` or in the `BasicMessageComposer` component?

**Pull request merged**